src/ray/raylet/worker_pool.cc

```diff
@@ -992,21 +992,32 @@ std::shared_ptr<WorkerInterface> WorkerPool::PopWorker(
       // Start a new worker process.
 
       if (task_spec.HasRuntimeEnv()) {
-        // create runtime env.
-        agent_manager_->CreateRuntimeEnv(
-            task_spec.JobId(), task_spec.SerializedRuntimeEnv(),
-            [this, start_worker_process_fn, &state, task_spec, runtime_env_hash](
-                bool successful, const std::string &serialized_runtime_env_context) {
-              if (successful) {
-                start_worker_process_fn(task_spec, state, {}, false, runtime_env_hash,
-                                        task_spec.SerializedRuntimeEnv(),
-                                        serialized_runtime_env_context);
-              } else {
-                RAY_LOG(ERROR) << "Creating runtime environment failed. The "
-                                  "corresponding task will be failed.";
-                runtime_env_setup_failed_callback_(task_spec.TaskId());
-              }
-            });
+        // create runtime env unless this env is already pending creation.
+        auto it = pending_runtime_envs_.find(runtime_env_hash);
+        if (it == pending_runtime_envs_.end()) {
+          pending_runtime_envs_.insert(runtime_env_hash);
+          agent_manager_->CreateRuntimeEnv(
+              task_spec.JobId(), task_spec.SerializedRuntimeEnv(),
+              [this, start_worker_process_fn, &state, task_spec, runtime_env_hash](
+                  bool successful, const std::string &serialized_runtime_env_context) {
+                pending_runtime_envs_.erase(runtime_env_hash);
```

Comment:
@edoakes this approach with `pending_runtime_envs_` looks different from the approach we discussed with the map to states (`PENDING`, `DONE`) but I think the outcome is the same--just wanted to double check that I wasn't overlooking something.

Reply:
I think the behavior here is slightly different than what we discussed previously. With the current implementation, we're effectively throttling it so that we can only ever have a single outstanding RPC to the agent. In the case where we've already created the env, we might want to start up a number of workers quickly and with this implementation we'll be doing it serially.

This behavior might be OK, but it would be a different from what we have with the non-runtime env case and I'd prefer to keep it the same if possible.

Reply:
Ah good point, let me go back to the other approach.